core/vm/ipgraph_with_kind.go

```diff
@@ -0,0 +1,365 @@
+package vm
+
+import (
+	"bytes"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+var (
+	royaltyPolicyKindLAP            = big.NewInt(0)
+	royaltyPolicyKindLRP            = big.NewInt(1)
+	setRoyaltyWithKindSelector      = crypto.Keccak256Hash([]byte("setRoyalty(address,address,uint256,uint256)")).Bytes()[:4]
+	getRoyaltyWithKindSelector      = crypto.Keccak256Hash([]byte("getRoyalty(address,address,uint256)")).Bytes()[:4]
+	getRoyaltyStackWithKindSelector = crypto.Keccak256Hash([]byte("getRoyaltyStack(address,uint256)")).Bytes()[:4]
+)
+
+type ipGraphWithPolicyKind struct {
+	ipGraph
+}
+
+func (c *ipGraphWithPolicyKind) RequiredGas(input []byte) uint64 {
+	if len(input) < 4 {
+		return 0
+	}
+	selector := input[:4]
+	switch {
+	case bytes.Equal(selector, addParentIpSelector):
+		return c.ipGraph.RequiredGas(input)
+	case bytes.Equal(selector, hasParentIpSelector):
+		return c.ipGraph.RequiredGas(input)
+	case bytes.Equal(selector, getParentIpsSelector):
+		return c.ipGraph.RequiredGas(input)
+	case bytes.Equal(selector, getParentIpsCountSelector):
+		return c.ipGraph.RequiredGas(input)
+	case bytes.Equal(selector, getAncestorIpsSelector):
+		return c.ipGraph.RequiredGas(input)
+	case bytes.Equal(selector, getAncestorIpsCountSelector):
+		return c.ipGraph.RequiredGas(input)
+	case bytes.Equal(selector, hasAncestorIpsSelector):
+		return c.ipGraph.RequiredGas(input)
+	case bytes.Equal(selector, setRoyaltyWithKindSelector):
+		return ipGraphWriteGas
+	case bytes.Equal(selector, getRoyaltyWithKindSelector):
+		royaltyPolicyKind := new(big.Int).SetBytes(getData(input, 64, 32))
+		if royaltyPolicyKind.Cmp(royaltyPolicyKindLAP) == 0 {
+			return ipGraphReadGas * (averageAncestorIpCount * 3)
```

Comment:
What is the thinking in the several gas cost estimations to use "average" vs a concrete number?

Reply:
When performing estimating gas cost for precompile, it won't access state so it cannot get the concrete/actual numbers.